mods/classic-taskbar-background-fix.wh.cpp

```diff
@@ -0,0 +1,368 @@
+
+    //copy the existing content from hdc
+    BitBlt(memDC, 0, 0, rect.right - rect.left, rect.bottom - rect.top, hdc, rect.left, rect.top, SRCCOPY);
+
+    //get pixels array from bitmap
+    BITMAPINFO bmi = {};
+    bmi.bmiHeader.biSize = sizeof(BITMAPINFOHEADER);
+    bmi.bmiHeader.biWidth = rect.right - rect.left;
+    bmi.bmiHeader.biHeight = -(rect.bottom - rect.top);     //negative height to ensure top-down bitmap
+    bmi.bmiHeader.biPlanes = 1;
+    bmi.bmiHeader.biBitCount = 32;  //32-bit color depth
+    bmi.bmiHeader.biCompression = BI_RGB;
+
+    int pixelCount = (rect.right - rect.left) * (rect.bottom - rect.top);
+    COLORREF* pixels = new COLORREF[pixelCount];
+    GetDIBits(memDC, memBitmap, 0, rect.bottom - rect.top, pixels, &bmi, DIB_RGB_COLORS);
+
+    //modify the pixels
+    for (int i = 0; i < pixelCount; ++i) {
+        if (pixels[i] == oldColor)
+            pixels[i] = newColor;
+    }
+
+    //save pixels array back to bitmap
+    SetDIBits(memDC, memBitmap, 0, rect.bottom - rect.top, pixels, &bmi, DIB_RGB_COLORS);
+    delete[] pixels;
+
+    //blit the modified content back to hdc
+    BitBlt(hdc, rect.left, rect.top, rect.right - rect.left, rect.bottom - rect.top, memDC, 0, 0, SRCCOPY);
+
+    //clean up
+    SelectObject(memDC, oldBitmap);
+    DeleteObject(memBitmap);
+    DeleteDC(memDC);
+}
+
+BOOL WINAPI EndPaintHook(
+    IN HWND                 hWnd,
+    IN const PAINTSTRUCT*   lpPaint
+) {
+    if (lpPaint && lpPaint->hdc && WindowNeedsBackgroundRepaint(hWnd)) {
+
+        COLORREF black = RGB(0, 0, 0);
+        COLORREF buttonFace = GetSysColor(COLOR_3DFACE);
+        if (buttonFace != black)
+            ConditionalFillRect(lpPaint->hdc, lpPaint->rcPaint, black, buttonFace);
+    }
+
+    return pOriginalEndPaint(hWnd, lpPaint);
+}
+
+
+void TriggerTaskbarRepaint(HWND hwndTaskbar) {
+
+    if (!hwndTaskbar) {     //is taskbar hwnd already detected?
+        hwndTaskbar = FindWindowW(L"Shell_TrayWnd", NULL);
+        Wh_Log(L"hwndTaskbar: 0x%llX", (long long)hwndTaskbar);
+    }
+
+    if (hwndTaskbar)
+        RedrawWindow(hwndTaskbar, NULL, NULL, RDW_INVALIDATE | RDW_ALLCHILDREN);
+}
+
+bool TryInit(bool* abort, bool callApplyHookOperations) {
+
+    HWND hwndTaskbar = FindWindowW(L"Shell_TrayWnd", NULL);
+    Wh_Log(L"hwndTaskbar: 0x%llX", (long long)hwndTaskbar);
+    if (!hwndTaskbar)
+        return false;   //retry
+
+    DWORD taskbarProcessId;
+    if (!GetWindowThreadProcessId(hwndTaskbar, &taskbarProcessId))
+        return false;   //retry
+
+    Wh_Log(L"taskbarProcessId: %u", taskbarProcessId);
+    if (taskbarProcessId != GetCurrentProcessId()) {
+        Wh_Log(L"Not a taskbar process, not hooking");
+        *abort = true;
+        return false;
+    }
+
+
+    Wh_Log(L"Initialising hooks...");
+
+
+    HMODULE hUser32 = GetModuleHandle(L"user32.dll"); 	//when we reach here then we can be sure that user32.dll has been already loaded

```

Comment:
What you did works, but I'm pretty sure user32.dll is available from the very beginning, so you can just set the hooks in `Wh_ModInit`. Avoiding `Wh_ApplyHookOperations` is also better for performance.

Reply:
I guess this question depends mostly on how to solve the next question. Right now there is a small issue with hooking CreateWindowEx, described below.

Reply:
Alternatively, yes, I could set the hook in Wh_ModInit. And then use the thread for detecting whether the hook function should actually do anything besides passing on to the original function.